### ISSUE TITLE
Scipy Optimize Lab

### ISSUE THREAD

**Issue by davidreber** [Foundations-of-Applied-Mathematics/Labs#1140]
- [ ] Write a better objective statement
- [ ] Fix problem 1
- [ ] Replace problem 4 with great applications!

**Comment by davidreber** [comment]
The test pull request is at #1139. What does @shanemcq18 think?

**Comment by shanemcq18** [comment]
Check out #1124 for an example.

**Comment by davidreber** [comment]
Note: by not doing the Newton-CG method in Problem 1, we eliminate students having to add options about whether to include the hessian or not...I think that's a good thing, because while that's useful, it seems rather problem-specific. What do you think @shanemcq18?

**Comment by shanemcq18** [comment]
@davidreber It could be constructive to have students use the hessian at least once, or at least realize that some of these algorithms benefit from having the hessian (or at least the jacobian) provided. Put Newton-CG in for now; if we don't like how it turns out we can rip it out later.

**Comment by shanemcq18** [comment]
@davidreber I'm really liking the `blackbox()` problem. Very cool look ahead to some Volume 4 concepts. I found that BFGS was actually better at doing the minimum than Powell, which is good since I don't think we mention Powell in the lab. Should this lab be turned in as an IPython Notebook (.ipynb) or as a plain Python file (.py)? If we do a notebook, that might facilitate pictures and written explanation.

I still really like the idea of plotting a 3D surface, finding the minimum, and plotting the surface and the minimal point together. We don't quite do that yet in this lab. Thoughts on where that could go in? Maybe as part of the basin hopping problem?

**Comment by davidreber** [comment]
@shanemcq18 In #1171, I made your changes to the `blackbox()` problem more consistent; I like that it uses BFGS.

I think turning it in as an IPython Notebook is a good idea. As for the 3D surface, we already do that in this lab. If you run Problem 3 in the solutions file, you'll find a wonderful just-transparent-enough 3D plot with the minimum for the basin hopping problem that I'm quite proud of.

**Comment by jmorrise** [comment]
@davidreber: @shanemcq18 said you could use feedback on the labs you've been working on. I'm new to this, so is there anything one thing in this lab that you want me to look at?

At a first pass: problem 1 is a good basic starting problem. 
I like the idea of minimizing a blackbox function in problem 2, and agree with Shane that it fits well with volume 4. Cool cool. It would be neat to have the minimizer be more complex than a straight line - thoughts?
Problem 3 is good although I think the explanation of basinhopping before the problem could be simplified/shortened. The 3d graph is cool (might look nicer in polar instead of rectangular, as in the figure). 
Still getting to problems 4 and 5. Root finding and curve fitting both seem like important topics to cover. The description of convection before problem 5 strikes me as a little lengthy. 

Let me know what specifically you'd like feedback on; if you just want any and all feedback, or none, that works too.

**Comment by davidreber** [comment]
@jmorrise Yeah, while I'd like your feedback on the entire lab, I am particularly interested in getting feedback on the unchecked items in the issue list, under the Proofreading section.

Along with that, I do think having a more interesting minimizer for the blackbox problem would be nice, but when I was considering other options, I was concerned the problem was becoming too difficult. But now that you mention it, I guess that wouldn't make it any more difficult for the students - so I'll go ahead and take a look at that. Can you tell me if there's any ways to simplify the problem for the students? (I want to make sure any difficulties for the students are teaching them too, not just getting in the way of their learning).

How would you simplify/shorten the basinhopping problem? I need help thinking of more ideas - I tried a few, but in the end they weren't really much better.  

Besides that, give me all the feedback you've got! Thanks.

**Comment by jmorrise** [comment]
**Feedback 1**
- General:
    - Where is plots.py? There are Python files to create some of the plots, but not all of them.
- Problem 1:
    - Plot the minimizer on the graph of the Rosenbrock function
- Problem 2:
    - I don't think the problem needs to be simplified, although it could be described more clearly (see below).
    - I don't think the section and problem description really stresses the concept of a black box function: "Here's some function, you don't need to know the source code or how it works, just minimize it." I could see students getting confused, opening up `blackbox_module` and trying to figure out what to do with it. Restructuring the problem statement might make it clearer. Thoughts?
    - Possibly change the minimizer to something other than a straight line by editing `blackbox()`. Could make it a curve by returning a weighted sum of the line segments. Something like this at the end for example:
    ```
weights = np.linspace(1,3,19)
return np.sum(partial_norms*weights)
    ```
    - Could include a figure showing sample output. Maybe not needed.
    - `opt.minimize()` has a cool optional argument `callback` which can be used to store the intermediate results. This could be used to visualize the progress of the algorithm towards the minimizer (maybe just in a visual in the lab and not by the students. rough example attached)
- Problem 3:
    - Your edits to the explanation are good, much simpler. The problem is good.
    - I think the arrows `>>>` in the plotting code should be removed, since people will copy/paste it and the arrows will just annoy them.
- Problem 4:
    - Seems a little short on explanation, but it isn't a very complicated idea. At least comment the code example to describe what is happening.
- Problem 5:
    - Same as above - code is a little short on explanation. Maybe explain the inputs to `curve_fit`: `func` takes `x` as its first argument and outputs `y`, and `curve_fit` tries to estimate the rest of the arguments. Would be nice to have a sentence saying what the example is doing.
- Will look at the rest of the problem later.
Blackbox example: ![blackbox](https://cloud.githubusercontent.com/assets/14076456/16888752/e59ea90e-4a96-11e6-9ba6-a0c1407a444c.png)
